include/revng/Pipeline/LibraryRegistry.h

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <utility>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/Pipeline/Loader.h"
+
+namespace Pipeline {
+
+/// a library registry is the general way of registering
+/// factories and kinds from down stream libraries.
+///
+/// it is possible that a factory my require to look up
+/// something into the builder and thus this class used to that purpose
+///
+/// for those kinds and containers and pipes that do not require to look inside
+/// the loader look further into this file to fine more declarative classes.
+class LibraryRegistry {
```

Comment:
from the outside what happens is that one uses the class LibraryRegistry only when he wants to do something like `LibraryRegistry::RegisterAllSomething()`

while when one to register one of its classes he ends up doing `RegisterX<MyClass> ()` by using one of the derived classes

i can split the two side of this behavior into two classes but it only introduces the necessity of forward declaring stuff and adding friend classes, so i opted for this

Reply:
I see, but it's called `LibraryRegistry` but it doesn't keep track of any library.
What if simply call it `Registry` (i.e., `Pipeline::Registry`)?

Reply:
ahh i see, yeah it make sense

Reply:
yeah i can't use default arguments because they would depend on a method of another class which depends itself on this header.

i dropepd the various = default toh, since they were redundant